### DISCUSSION TITLE
UI Dislplay Energie in Colums i.e on a Shelly 2.5 looks messy

### DISCUSSION THREAD

**Discussion by Mark-Muc** [arendst/Tasmota#15342] [Q&A] [answered]
Since
```
"[11.0.0.4] 20220402
Changed
	- Display of energy values in GUI use columns when define USE_ENERGY_COLUMN_GUI is enabled (default)"
```
the display on a dual channel measurement energy device (Shelly 2.5) looks destroyed in the UI webpage.

```
Temperatur	51.6 °C
Spannung		234	V
Frequenz		50	Hz
Strom		0.000	0.000	A
Leistung		0	0	W
Scheinleistung		0	0	VA
Blindleistung		0	0	VAr
Leistungsfaktor		0.00	0.00	
Energie heute		0.000	kWh
Energie gestern		0.000	kWh
Energie gesamt		0.000	kWh
```
Regards 
Mark

**Comment by sfromis** [comment, marked as answer]
In this situation a text copy/paste does not get across what you mean by "destroyed". Here's what I see on a Shelly 2.5, which does not look bad: 
![image](https://user-images.githubusercontent.com/47082390/162712458-14aaf489-9740-413a-8c02-eec480bfcbb4.png)

Same layout with a binary from 20220402 and from yesterday. I'm using the Chrome browser, if you get something messed up it could be a question of your browser. Which one are you using?

**Comment by Mark-Muc** [comment]
yes that's exactly it looks the same by me..
didn't know how to add a pic ..
try again to show what i mean:
![image](https://user-images.githubusercontent.com/67410847/162715019-dee44532-3c6e-4020-84ed-fa2582bf4a94.png)
and for my eyes it look a bit "destroyed"

Mark

**Reply by barbudor** [reply]
Measures that are different per channel are on a separated column, measures that are common are centered.
It may looks a bit strange but that makes it clear what is common and what is per-channel
Unless specified otherwise by the driver Voltage and Frequency are common to all channels.
And since 9.5.0.6, `SetOption129 1` allows Energy to be reported per-channel too.

**Reply by arendst** [reply]
The issue is caused by HTML table commands.

What you see here is that:
- All sensors provide their own table row layout. Until the multi column layout this was ALWAYS two columns; value(s) and units.
- Newer sensors can provide different values per channel/phase/relay resulting in a mess if you have four different values separated by a slash like x/x/x/x. They are left aligned within the column as otherwise if aligned on the decimal point the column would grow to wide if you have ie a current of 0.010A and a total power of 1234.002 kWh. This looks even more shitty on a phone. To make a HTML multi column work a new table statement is neede resultin in the new lines not knowing where the value/unit column width is.
- The first entry (temperature) fills a two column table row as all sensors currently do.
- The following entries are provided by (in this case) the Shelly energy driver. It supports two values per relay for most units except voltage and frequency. The new layout supports these columns and places single values centered over these columns. It wouldn't make sesnse to align them with the first row as in that case the voltage could be interpreted as only for the second relay).
- The last three entries can either be single ot multi column depending on ``SO129``.

In a perfect world the code would gather all sensor information BEFORE providing it to the user. This way correct column width can be calculated by the HTML table code. It would mean a tremendous lot of code and storage to get this done in time. No option for tasmota.

You have three options:
- Add more resolution to power (``wattres 2``) to increase reading
- Accept current sate.
- Compile your own disabling the new define.

![image](https://user-images.githubusercontent.com/11044339/162768395-72d91b98-e9b5-483f-9c52-b6bd065515ef.png)

**Reply by sfromis** [reply]
I still think that the cells with numbers would look "less wobbly" if simply right-aligned, as there will normally be at least 3 values with same number of decimals on top of each other.

**Reply by arendst** [reply]
I hear you (again) ;-)

It could be a good alternative especially if one only enables 2 decimals for all.....

I will add it to my list and see if it's easily done (with less or minimal extra code size).

**Reply by arendst** [reply]
How about this:

![image](https://user-images.githubusercontent.com/11044339/162786774-6a861188-f79a-4780-8284-ef77d428e5d6.png)

and this:

![image](https://user-images.githubusercontent.com/11044339/162786855-09947e5a-b9cc-40a6-94cd-bfab60c73cf0.png)

Implemented in latest commit.

Shoot.

**Reply by joba-1** [reply]
> Shoot.

I like it - will try it on the Dual R3.

But you know what kind of issues will come up the next days, right? How can I set the labels for L1-x? Why is time and other sensor data not aligned? :P

**Reply by sfromis** [reply]
Looks better now :-) 

Not having that time posted there (did not see much point in that), the main nitpick I have is that the temperature number and unit are not in the same column structure as voltage etc.

**Reply by arendst** [reply]
Regarding alignment see my explanation above. Any solution is welcome.

**Reply by sfromis** [reply]
I'm fine with the alignment differences I get if selecting different number of decimals for different value types. It's just that the whole table could be with the unit (if present, including `°C`) in the right-hand column, and the value in the table column left of that. Making everything 2 or 3 decimals is then a question of how bad an OCD the user has ;-) I get no tics when having energy with 5 decimals, and power with 3.

**Reply by sfromis** [reply]
Of course, I'm aware that the temperature sensor is outside of the rendering done by the energy driver, and thus integration may not be straightforward.

**Reply by sfromis** [reply]
If you (like me) prefer the energy layout with right-aligned columns, with latest binary you should also use `SetOption138 1`

Unfortunately, it does not apply to other sensors :-D

**Reply by arendst** [reply]
Correct. As we're close to release I didn't want to make too much changes.

I do like to horizontal line though making up for the non-alignment of non-energy items.

![image](https://user-images.githubusercontent.com/11044339/162995525-a6cec4f9-6652-478f-be2e-f3ee9a6882b3.png)

**Reply by sfromis** [reply]
I like the `<hr>` too.

**Comment by Mark-Muc** [comment]
first of all i have to apologise for starting such a big discussion .. and thank you, to take care about the "problem".
I now set the so129 1 .. now it looks much better and it's no more hustle @ all!! Yesterday a saw 2 pic's of a proposal how to display .. both looked much better, then the current view!!
anyhow "so129 1" does the trick!!
Again THX
Mark

**Comment by Mark-Muc** [comment]
Me too ..ggg
I try to built a version today and place a pic in the tread!!
Thy again!!
Mark

**Comment by Mark-Muc** [comment]
built the new version and with
`SetOption138 1`
right alligned 
![image](https://user-images.githubusercontent.com/67410847/163141248-a5d8757b-f7f2-4ad8-a001-636b6bd2a9d4.png)

looks MUCH MUCH better ..
Thanks
Mark
